ext/standard/string.c

```diff
@@ -4801,7 +4801,12 @@ PHP_FUNCTION(setlocale)
 					if (BG(ctype_string)) {
 						zend_string_release_ex(BG(ctype_string), 0);
 					}
-					if (len == ZSTR_LEN(loc) && !memcmp(ZSTR_VAL(loc), retval, len)) {
+					if (len == 1 && *retval == 'C') {
```

Comment:
I'm fairly certain that `retval` was the **original locale**, not the **new locale**.

**This should check ZSTR_VAL(loc) instead, e.g. `zend_string_equals_literal(locale, "C")`**

Reply:
setlocale() returns the locale that was set, not the previous one. This is necessary because it may not be the same as what was passed to the setlocale() call, e.g. when using `""` environment locale.

Reply:
Ok, good. I was mixed up with other C and PHP APIs that did that - the c code is returning the new locale

https://en.cppreference.com/w/c/locale/setlocale
https://www.php.net/manual/en/function.setlocale.php#refsect1-function.setlocale-returnvalues